src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -109,7 +126,10 @@ public static ConsoleKeyInfo ReadKey(bool intercept)
                 keyInfo = new ConsoleKeyInfo('\r', keyInfo.Key, shift, alt, control);
             }
 
-            if (!intercept && !previouslyProcessed) Console.Write(keyInfo.KeyChar);
+            if (!intercept && !previouslyProcessed && keyInfo.KeyChar != '\0')
```

Comment:
@stephentoub @wtgodbe I've added `!= '\0` check here. This isn't related to the cursor caching. The `StdInReader.ReadLine` method also doesn't write these, and I think it is safe to filter those out here too.

Reply:
What does this help?

Reply:
I hadn't noticed the question. It stops from doing a single byte write with '\0' that is ignored by the terminal.

Reply:
Sure, thanks. But what causes the null in the first place? Is it actually common?

Reply:
For example, if you press an arrow key, you get a '\0' char.

Reply:
Ah, got it.